pkg/operator/operator.go

```diff
@@ -743,3 +757,67 @@ func (o *Operator) Config(ctx context.Context, key string) (*manifests.Config, e
 
 	return c, nil
 }
+
+func (o *Operator) Upgradeable(ctx context.Context, message *string, reason *string) (configv1.ConditionStatus, error) {
+	if o.lastKnowInfrastructureConfig.HighlyAvailableInfrastructure() {
+		workloadsCorrectlySpread, err := o.WorkloadsCorrectlySpread(ctx, message, reason)
+		if err != nil {
+			return configv1.ConditionUnknown, err
+		}
+		if !workloadsCorrectlySpread {
+			return configv1.ConditionFalse, nil
+		}
+	}
+	return configv1.ConditionTrue, nil
+}
+
+func (o *Operator) workloadCorrectlySpread(ctx context.Context, namespace string, sel map[string]string) (bool, error) {
+	podList, err := o.client.ListPods(ctx, namespace, metav1.ListOptions{LabelSelector: labels.FormatLabels(sel)})
+	if err != nil {
+		return false, err
+	}
+
+	// Skip the check if we can't get enough pods. This prevents setting the status when the cluster is degraded.
+	if len(podList.Items) <= 1 {
+		return true, nil
+	}
+
+	nodes := make(map[string]struct{}, len(podList.Items))
+	for _, pod := range podList.Items {
+		nodes[pod.Spec.NodeName] = struct{}{}
+	}
+
+	return len(nodes) > 1, nil
+}
+
+func (o *Operator) WorkloadsCorrectlySpread(ctx context.Context, message *string, reason *string) (bool, error) {
+	type workload struct {
+		namespace     string
+		name          string
+		labelSelector map[string]string
+	}
+
+	workloads := []workload{
+		{o.namespace, "prometheus-k8s", map[string]string{"app.kubernetes.io/name": "prometheus"}},
+		{o.namespace, "alertmanager-main", map[string]string{"app.kubernetes.io/name": "alertmanager"}},
+	}
+	if o.userWorkloadEnabled {
+		// TODO: Update thanos-ruler label selector to `app.kubernetes.io/name` once updated in the CRD.
+		workloads = append(workloads, workload{o.namespaceUserWorkload, "user-workload", map[string]string{"thanosRulerName": "user-workload"}})
```

Comment:
we should consider updating this in https://github.com/openshift/cluster-monitoring-operator/pull/1205